test-js/unit/pentaho/lang/Filter.Spec.js

```diff
@@ -14,61 +14,297 @@
+        var filteredData = combination.filter(data);
+
+        expect(filteredData.getNumberOfRows()).toBe(3);
+        expect(filteredData.getValue(0, 0)).toBe("A");
+        expect(filteredData.getValue(1, 0)).toBe("B");
+        expect(filteredData.getValue(2, 0)).toBe("C");
+      });
+
+    });
+
+    describe("negation ", function() {
+      it("should return the negation of a simple filter", function() {
+        var sales12k = new Filter.IsEqual("sales", 12000);
+
+        var combination = sales12k.negation(sales12k);
+        var filteredData = combination.filter(data);
+
+        expect(filteredData.getNumberOfRows()).toBe(5);
+        [
+          "B", "D", "E", "F", "G"
+        ].forEach(function(product, idx) {
+          expect(filteredData.getValue(idx, 0)).toBe(product);
+        });
+
+      });
+
+      it("should return the negation of a simple filter", function() {
+        var sales12k = Filter.negation(Filter.isEqual("sales", 12000));
+
+        var filteredData = sales12k.filter(data);
+
+        expect(filteredData.getNumberOfRows()).toBe(5);
+        [
+          "B", "D", "E", "F", "G"
+        ].forEach(function(product, idx) {
+          expect(filteredData.getValue(idx, 0)).toBe(product);
+        });
+
+      });
+
+      it("should return the negation of a complex filter", function() {
+        var filter = new Filter.Or();
+        expect(filter.type).toBe("or");
+        expect(filter.predicate()).toBe(false);
+        filter
+            .insert(new Filter.IsEqual("product", "A"))
+            .insert(new Filter.IsEqual("product", "B"))
+            .insert(new Filter.IsEqual("product", "C"));
+
+        var combination = filter.negation(filter);
+        var filteredData = combination.filter(data);
+
+        expect(filteredData.getNumberOfRows()).toBe(4);
+        [
+          "D", "E", "F", "G"
+        ].forEach(function(product, idx) {
+          expect(filteredData.getValue(idx, 0)).toBe(product);
+        });
+
+      });
+
+      it("should return the negation of a complex filter", function() {
+        var filter = Filter.or();
+        expect(filter.type).toBe("or");
+        expect(filter.predicate()).toBe(false);
+        filter
+            .insert(Filter.isEqual("product", "A"))
+            .insert(Filter.isEqual("product", "B"))
+            .insert(Filter.isEqual("product", "C"));
+
+        var combination = Filter.negation(filter);
+        var filteredData = combination.filter(data);
+
+        expect(filteredData.getNumberOfRows()).toBe(4);
+        [
+          "D", "E", "F", "G"
+        ].forEach(function(product, idx) {
+          expect(filteredData.getValue(idx, 0)).toBe(product);
+        });
+
+      });
+    });
+
+    describe("contains ", function() {
+
+      it("should return the contains of a simple filter", function() {
+        var productFilter = new Filter.IsIn("product", ["A", "B", "C", "D", "E", "F", "G"]);
+        var filteredData = productFilter.filter(data);
```

Comment:
why bother adding the line `var filteredData = productFilter.filter(data);` if it's not being used?